### DISCUSSION TITLE
Binaries created with LIBIGL_USE_STATIC_LIBRARY ON and OFF link to the same libraries but different file size

### DISCUSSION THREAD

**Discussion by ansonl** [libigl/libigl#2297] [General]
Hi,

When I compile 2 versions of a sample program using libigl using `LIBIGL_USE_STATIC_LIBRARY`=ON in the first version and `LIBIGL_USE_STATIC_LIBRARY`=OFF in the second version, the binary compiled with static=ON is larger in size than the binary compiled with static=OFF. Using `readelf -d` says the static version of the binary has a larger LOAD/.text section so I assumed that meant that all of libigl's dependencies were all packed into a static `libigl.a` before putting it into the static compiled output binary and the extra code from libigl's dependencies are in the larger LOAD/.text section. 

@alecjacobson For reference, I compiled the `meshboolean` example at https://github.com/alecjacobson/gp-cli with `LIBIGL_USE_STATIC_LIBRARY` set to `ON` and `OFF`. Both binaries run successfully -- however, running `ldd` on both binaries show the same libraries being pulled in at runtime. 

```
$ ldd meshboolean-shared-libigl
        linux-vdso.so.1 =>  (0x00007fff26fd9000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fb549924000)
        libstdc++.so.6 => /share/spack/gcc-10.3.0/gcc-12.2.0-4dp/lib64/libstdc++.so.6 (0x00007fb54950e000)
        libm.so.6 => /lib64/libm.so.6 (0x00007fb54920c000)
        libgcc_s.so.1 => /share/spack/gcc-10.3.0/gcc-12.2.0-4dp/lib64/libgcc_s.so.1 (0x00007fb548fee000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fb548c20000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fb549b40000)
```
```
$ ldd meshboolean-static-libigl
        linux-vdso.so.1 =>  (0x00007ffdbab95000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f3f17736000)
        libstdc++.so.6 => /share/spack/gcc-10.3.0/gcc-12.2.0-4dp/lib64/libstdc++.so.6 (0x00007f3f17320000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f3f1701e000)
        libgcc_s.so.1 => /share/spack/gcc-10.3.0/gcc-12.2.0-4dp/lib64/libgcc_s.so.1 (0x00007f3f16e00000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f3f16a32000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f3f17952000)
```

Due to the above, I'm confused about `LIBIGL_USE_STATIC_LIBRARY`'s usage. 

Why do both the shared and static compiled binaries have the same library imports? 

Shouldn't there be a `libigl.so` call for the shared version (because only the needed code of libigl.a is bundled into the static version)? If the same library import is expected, what is in the extra (unneeded?) code for the static version even though both binaries run fine on their own and call the exact same libraries?

Does `LIBIGL_USE_STATIC_LIBRARY=YES` not bundle all the program's dependencies into the binary at compilation and negate the need for the program to call additional libraries? 

My current assumption is that the static version has duplicate code of the libraries pulled in in the `ldd` output. But only the libigl portion of the code uses the extra code while the rest of the program somehow still uses the shared libraries on the system. I don't think my understanding is correct. 

Can anyone please advise on what is the correct understanding of `LIBIGL_USE_STATIC_LIBRARY`? The reason I'm asking is because I would like to understand creating a portable binary using libigl. Thanks.

**Comment by alecjacobson** [comment]
> Why do both the shared and static compiled binaries have the same library imports?
> Shouldn't there be a libigl.so call for the shared version (because only the needed code of libigl.a is bundled into the static version)? If the same library import is expected, what is in the extra (unneeded?) code for the static version even though both binaries run fine on their own and call the exact same libraries?

We're distinguishing between using libigl as 1) a header-only library and 2) as a static library of precompiled template instantiations. We don't maintain support for compiling a dynamic/shared library, though I believe others have done this in the past.

> Does LIBIGL_USE_STATIC_LIBRARY=YES not bundle all the program's dependencies into the binary at compilation and negate the need for the program to call additional libraries?

libigl.a will not contain binaries for dependencies. The core library only depends on Eigen (header-only) and stl.

**Reply by ansonl** [reply]
Oh, I see so libigl is only supported being used as a folder of imports (with headers) or as a static library libigl.a by the compiler. 

Then in this case of compiling the meshboolean program from gp-cli with `LIBIGL_USE_STATIC_LIBRARY=YES`, what is the origin of the extra code that is included? The shared binary is about 2mb and the static binary is 7mb so the static one can't fit the entire libigl.a which is 21mb. 

I still do not understand if the extra code is from portions of libigl.a containing actual libigl functional code needed for mesh boolean (not any libigl deps like eigen/stl/cgal) or something else that the program needs but doesn't import in the same binary compiled with the shared libigl (in this case is this code considered unneeded)? The shared libigl runs fine on a system without eigen lib in the library path so I'm wondering how that is possible if the shared binary doesn't include deps? 

Thanks.
